File: 845_longest_mountain_in_array.java
class Solution {
    // Time Complexity: O(len(A)), Space Complexity: O(1)
    public int longestMountain(int[] A) {
        int len = A.length;
        int ans = 0, base = 0;

        while (base < len) {
            int end = base;
            // 1. need to check whether end is left-boundary
            if (end + 1 < len && A[end] < A[end + 1]) {
                // 2. keep find the peak
                while (end + 1 < len && A[end] < A[end + 1]) {
                    end++;
                }

                // 3. check if it's the peak
                if (end + 1 < len && A[end] > A[end + 1]) {

                    // 4. find potential valley
                    while (end + 1 < len && A[end] > A[end + 1]) {
                        end++;
                    }
                    ans = Math.max(ans, end - base + 1);
                }
            }

            // update base position
            base = Math.max(end, base + 1);
        }
        return ans;
    }
}

File: python/845_longest_mountain_in_array.py
class Solution:
    def longestMountain(self, A: List[int]) -> int:
        base = 0
        ans = 0

        while base < len(A):
            end = base

            # check if is the left-boundary
            if end + 1 < len(A) and A[end] < A[end + 1]:
                # keep find the peak
                while end + 1 < len(A) and A[end] < A[end + 1]:
                    end += 1

                # check if is the right boundary
                if end + 1 < len(A) and A[end] > A[end + 1]:
                    # keep finding the valley
                    while end + 1 < len(A) and A[end] > A[end + 1]:
                        end += 1
                    ans = max(ans, end - base + 1)

            base = max(end, base + 1)
        return ans
